### PR TITLE
Improve interface check by method parameters

### PR DESCRIPTION
**Pull request by ksss**
Regarding checking interfaces, in addition to checking with `#respond_to?`, I believe that comparing `Method#parameters` with types can provide a more accurate interface check.

As an implementation, I've made it so that a match only occurs if the method has exactly the same arguments as the type notation. 
Does this match with the philosophy of RBS? 
For example, the following code actually works. 

```rbs
# sample.rbs
class Sample
  def foo: (a: Integer, b: String) -> void
end
```

```rb
# sample.rb
class Sample
  def foo(a = {})
    p a # { :a => 1, :b => 'b' }
  end
end

Sample.new.foo(a: 1, b: 'b')
```

However, for instance, steep would report `Ruby::MethodArityMismatch` and `Ruby::DifferentMethodParameterKind`. 
What kind of checks should be attempted with RBS alone?

### CODE REVIEW

**Comment by soutaro** [comment]
I think the parameter checking should be permissive. Having an optional argument in implementation while required argument in RBS should be allowed for example.

**Comment by soutaro** [comment]
Steep validates the method implementation and reports `DifferentMethodParameterKind`. But I don't think RBS should do it on the implementation side.

**Comment by ksss** [comment]
Thank you!
I think it seems good to use the criterion "if it can be called, it's OK"

**Comment by ksss** [comment]
I have reimplemented code to allow for many method call patterns.
I also added an ArgumentChecker to generate arguments from RBS and test whether a method can actually be called.

File: test/rbs/test/type_check_test.rb [hunk 1]
```diff
@@ -69,6 +69,197 @@ def to_int: () -> Integer
...
+          failure = Object.new
+          definer.call(failure)
+
+          no = ArgumentChecker.new(builder: builder, interface: interface).no_argument_error?(:f) do |args, kwargs, block|
+            failure.f(*args, **kwargs, &block)
+          end
+          refute no
+          refute typecheck.value(failure, parse_type(interface))
+        rescue
+          p [:failure, definer, interface]
```
Comment (soutaro) [inline test/rbs/test/type_check_test.rb:256]: Is this intentionally print using `p`? (Not sure if it should be.)

**Review by soutaro — APPROVED** [review]
Didn't review the implementation very carefully, but give it a try! 💪

File: test/rbs/test/type_check_test.rb (hunk 1 above)
Comment (ksss) [reply test/rbs/test/type_check_test.rb:256]: When a test failed, it was intentionally included to indicate which line of the case had failed.
However, I agree that it is complicated and will remove it.

**Pull request merged**